<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4 Getting the Summary Statistics | Comparison of PGS Generative Methods</title>
  <meta name="description" content="A deep dive into all the scripts and underlying though processes that create polygenic risk scores" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="4 Getting the Summary Statistics | Comparison of PGS Generative Methods" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A deep dive into all the scripts and underlying though processes that create polygenic risk scores" />
  <meta name="github-repo" content="pgs_book" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 Getting the Summary Statistics | Comparison of PGS Generative Methods" />
  
  <meta name="twitter:description" content="A deep dive into all the scripts and underlying though processes that create polygenic risk scores" />
  

<meta name="author" content="Scott Kulm" />


<meta name="date" content="2020-09-15" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="summary-statistics.html"/>
<link rel="next" href="adjusting-summary-statistics.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#directory-framework"><i class="fa fa-check"></i><b>1.1</b> Directory Framework</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>2</b> Quality Control</a><ul>
<li class="chapter" data-level="2.1" data-path="quality-control.html"><a href="quality-control.html#previously-applied-qc"><i class="fa fa-check"></i><b>2.1</b> Previously Applied QC</a></li>
<li class="chapter" data-level="2.2" data-path="quality-control.html"><a href="quality-control.html#initial-qc"><i class="fa fa-check"></i><b>2.2</b> Initial QC</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="summary-statistics.html"><a href="summary-statistics.html"><i class="fa fa-check"></i><b>3</b> Summary Statistics</a></li>
<li class="chapter" data-level="4" data-path="getting-the-summary-statistics.html"><a href="getting-the-summary-statistics.html"><i class="fa fa-check"></i><b>4</b> Getting the Summary Statistics</a><ul>
<li class="chapter" data-level="4.1" data-path="getting-the-summary-statistics.html"><a href="getting-the-summary-statistics.html#prepare-variants"><i class="fa fa-check"></i><b>4.1</b> Prepare Variants</a></li>
<li class="chapter" data-level="4.2" data-path="getting-the-summary-statistics.html"><a href="getting-the-summary-statistics.html#standardize-the-summary-statistics"><i class="fa fa-check"></i><b>4.2</b> Standardize the Summary Statistics</a></li>
<li class="chapter" data-level="4.3" data-path="getting-the-summary-statistics.html"><a href="getting-the-summary-statistics.html#heritability-estimation"><i class="fa fa-check"></i><b>4.3</b> Heritability Estimation</a></li>
<li class="chapter" data-level="4.4" data-path="getting-the-summary-statistics.html"><a href="getting-the-summary-statistics.html#genetic-correlation-estimation"><i class="fa fa-check"></i><b>4.4</b> Genetic Correlation Estimation</a></li>
<li class="chapter" data-level="4.5" data-path="getting-the-summary-statistics.html"><a href="getting-the-summary-statistics.html#final-changes-and-remarks"><i class="fa fa-check"></i><b>4.5</b> Final Changes and Remarks</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html"><i class="fa fa-check"></i><b>5</b> Adjusting Summary Statistics</a><ul>
<li class="chapter" data-level="5.0.1" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#computational-framework"><i class="fa fa-check"></i><b>5.0.1</b> Computational Framework</a></li>
<li class="chapter" data-level="5.1" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#simple"><i class="fa fa-check"></i><b>5.1</b> Simple</a></li>
<li class="chapter" data-level="5.2" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#clumping"><i class="fa fa-check"></i><b>5.2</b> Clumping</a></li>
<li class="chapter" data-level="5.3" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#double-weight"><i class="fa fa-check"></i><b>5.3</b> Double Weight</a></li>
<li class="chapter" data-level="5.4" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#ldpred"><i class="fa fa-check"></i><b>5.4</b> LDpred</a></li>
<li class="chapter" data-level="5.5" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#ldpred2"><i class="fa fa-check"></i><b>5.5</b> LDpred2</a></li>
<li class="chapter" data-level="5.6" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#sblup"><i class="fa fa-check"></i><b>5.6</b> SBLUP</a></li>
<li class="chapter" data-level="5.7" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#smtpred"><i class="fa fa-check"></i><b>5.7</b> SMTpred</a></li>
<li class="chapter" data-level="5.8" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#prscs"><i class="fa fa-check"></i><b>5.8</b> prsCS</a></li>
<li class="chapter" data-level="5.9" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#lassosum"><i class="fa fa-check"></i><b>5.9</b> lassosum</a></li>
<li class="chapter" data-level="5.10" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#tweedie"><i class="fa fa-check"></i><b>5.10</b> Tweedie</a></li>
<li class="chapter" data-level="5.11" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#sbayesr"><i class="fa fa-check"></i><b>5.11</b> SBayesR</a></li>
<li class="chapter" data-level="5.12" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#jampred"><i class="fa fa-check"></i><b>5.12</b> JAMPred</a></li>
<li class="chapter" data-level="5.13" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#winners-curse-lasso"><i class="fa fa-check"></i><b>5.13</b> Winner’s Curse Lasso</a></li>
<li class="chapter" data-level="5.14" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#winners-curse-likelihood"><i class="fa fa-check"></i><b>5.14</b> Winner’s Curse Likelihood</a></li>
<li class="chapter" data-level="5.15" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#winners-curse-2d"><i class="fa fa-check"></i><b>5.15</b> Winner’s Curse 2D</a><ul>
<li class="chapter" data-level="5.15.1" data-path="adjusting-summary-statistics.html"><a href="adjusting-summary-statistics.html#additional-methods"><i class="fa fa-check"></i><b>5.15.1</b> Additional Methods</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="compiling-the-scores.html"><a href="compiling-the-scores.html"><i class="fa fa-check"></i><b>6</b> Compiling the Scores</a><ul>
<li class="chapter" data-level="6.1" data-path="compiling-the-scores.html"><a href="compiling-the-scores.html#simple-score-approach"><i class="fa fa-check"></i><b>6.1</b> Simple Score Approach</a><ul>
<li class="chapter" data-level="6.1.1" data-path="compiling-the-scores.html"><a href="compiling-the-scores.html#what-to-worry-about"><i class="fa fa-check"></i><b>6.1.1</b> What to Worry About</a></li>
<li class="chapter" data-level="6.1.2" data-path="compiling-the-scores.html"><a href="compiling-the-scores.html#controlling-and-assembling"><i class="fa fa-check"></i><b>6.1.2</b> Controlling and Assembling</a></li>
<li class="chapter" data-level="6.1.3" data-path="compiling-the-scores.html"><a href="compiling-the-scores.html#verifying-everything-worked"><i class="fa fa-check"></i><b>6.1.3</b> Verifying Everything Worked</a></li>
<li class="chapter" data-level="6.1.4" data-path="compiling-the-scores.html"><a href="compiling-the-scores.html#other-options"><i class="fa fa-check"></i><b>6.1.4</b> Other Options</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="score-analysis-set-up.html"><a href="score-analysis-set-up.html"><i class="fa fa-check"></i><b>7</b> Score Analysis Set-Up</a><ul>
<li class="chapter" data-level="7.1" data-path="score-analysis-set-up.html"><a href="score-analysis-set-up.html#defining-the-phenotypes"><i class="fa fa-check"></i><b>7.1</b> Defining the Phenotypes</a></li>
<li class="chapter" data-level="7.2" data-path="score-analysis-set-up.html"><a href="score-analysis-set-up.html#making-the-phenotype"><i class="fa fa-check"></i><b>7.2</b> Making the Phenotype</a></li>
<li class="chapter" data-level="7.3" data-path="score-analysis-set-up.html"><a href="score-analysis-set-up.html#other-scores"><i class="fa fa-check"></i><b>7.3</b> Other Scores</a></li>
<li class="chapter" data-level="7.4" data-path="score-analysis-set-up.html"><a href="score-analysis-set-up.html#other-covariates"><i class="fa fa-check"></i><b>7.4</b> Other Covariates</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="tuning.html"><a href="tuning.html"><i class="fa fa-check"></i><b>8</b> Tuning</a><ul>
<li class="chapter" data-level="8.1" data-path="tuning.html"><a href="tuning.html#generating-metrics"><i class="fa fa-check"></i><b>8.1</b> Generating Metrics</a><ul>
<li class="chapter" data-level="8.1.1" data-path="tuning.html"><a href="tuning.html#overall-set-up"><i class="fa fa-check"></i><b>8.1.1</b> Overall Set-Up</a></li>
<li class="chapter" data-level="8.1.2" data-path="tuning.html"><a href="tuning.html#survival-set-up"><i class="fa fa-check"></i><b>8.1.2</b> Survival Set-up</a></li>
<li class="chapter" data-level="8.1.3" data-path="tuning.html"><a href="tuning.html#survival-base-metrics"><i class="fa fa-check"></i><b>8.1.3</b> Survival Base Metrics</a></li>
<li class="chapter" data-level="8.1.4" data-path="tuning.html"><a href="tuning.html#survival-score-metrics"><i class="fa fa-check"></i><b>8.1.4</b> Survival Score Metrics</a></li>
<li class="chapter" data-level="8.1.5" data-path="tuning.html"><a href="tuning.html#logistic-regression-base-metrics"><i class="fa fa-check"></i><b>8.1.5</b> Logistic Regression Base Metrics</a></li>
<li class="chapter" data-level="8.1.6" data-path="tuning.html"><a href="tuning.html#logistic-regression-score-metrics"><i class="fa fa-check"></i><b>8.1.6</b> Logistic Regression Score Metrics</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="tuning.html"><a href="tuning.html#choosing-the-best-score"><i class="fa fa-check"></i><b>8.2</b> Choosing the Best Score</a></li>
<li class="chapter" data-level="8.3" data-path="tuning.html"><a href="tuning.html#plotting"><i class="fa fa-check"></i><b>8.3</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="non-predictive-evaluation.html"><a href="non-predictive-evaluation.html"><i class="fa fa-check"></i><b>9</b> Non-Predictive Evaluation</a><ul>
<li class="chapter" data-level="9.1" data-path="non-predictive-evaluation.html"><a href="non-predictive-evaluation.html#analysis-of-score-sizes"><i class="fa fa-check"></i><b>9.1</b> Analysis of Score Sizes</a></li>
<li class="chapter" data-level="9.2" data-path="non-predictive-evaluation.html"><a href="non-predictive-evaluation.html#analysis-of-phenotype-definitions"><i class="fa fa-check"></i><b>9.2</b> Analysis of Phenotype Definitions</a></li>
<li class="chapter" data-level="9.3" data-path="non-predictive-evaluation.html"><a href="non-predictive-evaluation.html#comparison-of-racial-groups"><i class="fa fa-check"></i><b>9.3</b> Comparison of Racial Groups</a></li>
<li class="chapter" data-level="9.4" data-path="non-predictive-evaluation.html"><a href="non-predictive-evaluation.html#comparison-of-siblings"><i class="fa fa-check"></i><b>9.4</b> Comparison of Siblings</a></li>
<li class="chapter" data-level="9.5" data-path="non-predictive-evaluation.html"><a href="non-predictive-evaluation.html#mathematical-comparison-of-score-methods"><i class="fa fa-check"></i><b>9.5</b> Mathematical Comparison of Score Methods</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="testing.html"><a href="testing.html"><i class="fa fa-check"></i><b>10</b> Testing</a></li>
<li class="chapter" data-level="11" data-path="predictive-evaluations.html"><a href="predictive-evaluations.html"><i class="fa fa-check"></i><b>11</b> Predictive Evaluations</a><ul>
<li class="chapter" data-level="11.1" data-path="predictive-evaluations.html"><a href="predictive-evaluations.html#expanded-covariates"><i class="fa fa-check"></i><b>11.1</b> Expanded Covariates</a></li>
<li class="chapter" data-level="11.2" data-path="predictive-evaluations.html"><a href="predictive-evaluations.html#other-polygenic-risk-scores"><i class="fa fa-check"></i><b>11.2</b> Other Polygenic Risk Scores</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Comparison of PGS Generative Methods</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="getting-the-summary-statistics" class="section level1">
<h1><span class="header-section-number">4</span> Getting the Summary Statistics</h1>
<p>Now we have to get the actual summary statistics. A key consideration is to pull from summary statistics with the largest possibly underlying sample size, because it will produce the most accurate estimates. The studies can be obtained with:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">#Files that are already haromized</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="fu">mkdir</span> Bentham</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="fu">wget</span> -O Bentham/raw_bentham.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/BenthamJ_26502338_GCST003156/harmonised/26502338-GCST003156-EFO_0002690-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="fu">mkdir</span> Christophersen</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="fu">wget</span> -O Christophersen/raw_christophersen.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/ChristophersenIE_28416818_GCST004296/harmonised/28416818-GCST004296-EFO_0000275-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-8" data-line-number="8"></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="fu">mkdir</span> Demenais</a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="fu">wget</span> -O Demenais/raw_demenais.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/DemenaisF_29273806_GCST005212/harmonised/29273806-GCST005212-EFO_0000270-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="fu">mkdir</span> Dubois</a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="fu">wget</span> -O Dubois/raw_dubois.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/DuboisPC_20190752_GCST000612/harmonised/20190752-GCST000612-EFO_0001060-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-14" data-line-number="14"></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="fu">mkdir</span> Liu-1</a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="fu">wget</span> -O Liu-1/raw_liu-1.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/LiuJZ_26192919_GCST003044/harmonised/26192919-GCST003044-EFO_0000384-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-17" data-line-number="17"></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="fu">mkdir</span> Liu-2</a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="fu">wget</span> -O Liu-2/raw_liu-2.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/LiuJZ_26192919_GCST003045/harmonised/26192919-GCST003045-EFO_0000729-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-20" data-line-number="20"></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="fu">mkdir</span> Malik</a>
<a class="sourceLine" id="cb3-22" data-line-number="22"><span class="fu">wget</span> -O Malik/raw_malik.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/MalikR_29531354_GCST005838/harmonised/29531354-GCST005838-EFO_0000712-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-23" data-line-number="23"></a>
<a class="sourceLine" id="cb3-24" data-line-number="24"><span class="fu">mkdir</span> Michailidou</a>
<a class="sourceLine" id="cb3-25" data-line-number="25"><span class="fu">wget</span> -O Michailidou/raw_michailidou.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/MichailidouK_29059683_GCST004988/harmonised/29059683-GCST004988-EFO_0000305-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-26" data-line-number="26"></a>
<a class="sourceLine" id="cb3-27" data-line-number="27"><span class="fu">mkdir</span> Nikpay</a>
<a class="sourceLine" id="cb3-28" data-line-number="28"><span class="fu">wget</span> -O Nikpay/raw_nikpay.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/NikpayM_26343387_GCST003116/harmonised/26343387-GCST003116-EFO_0000378-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-29" data-line-number="29"></a>
<a class="sourceLine" id="cb3-30" data-line-number="30"><span class="fu">mkdir</span> Okada</a>
<a class="sourceLine" id="cb3-31" data-line-number="31"><span class="fu">wget</span> -O Okada/raw_okada.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/OkadaY_24390342_GCST002318/harmonised/24390342-GCST002318-EFO_0000685-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-32" data-line-number="32"></a>
<a class="sourceLine" id="cb3-33" data-line-number="33"><span class="fu">mkdir</span> Onengut</a>
<a class="sourceLine" id="cb3-34" data-line-number="34"><span class="fu">wget</span> -O Onengut/raw_onengut.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/Onengut-GumuscuS_25751624_GCST005536/harmonised/25751624-GCST005536-EFO_0001359-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-35" data-line-number="35"></a>
<a class="sourceLine" id="cb3-36" data-line-number="36"><span class="fu">mkdir</span> Phelan</a>
<a class="sourceLine" id="cb3-37" data-line-number="37"><span class="fu">wget</span> -O Phelan/raw_phelan.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/PhelanCM_28346442_GCST004462/harmonised/28346442-GCST004462-EFO_0001075-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-38" data-line-number="38"></a>
<a class="sourceLine" id="cb3-39" data-line-number="39"><span class="fu">mkdir</span> Schumacher</a>
<a class="sourceLine" id="cb3-40" data-line-number="40"><span class="fu">wget</span> -O Schumacher/raw_schumacher.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/SchumacherFR_29892016_GCST006085/harmonised/29892016-GCST006085-EFO_0001663-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-41" data-line-number="41"></a>
<a class="sourceLine" id="cb3-42" data-line-number="42"><span class="fu">mkdir</span> Tsoi</a>
<a class="sourceLine" id="cb3-43" data-line-number="43"><span class="fu">wget</span> -O Tsoi/raw_tsoi.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/TsoiLC_23143594_GCST005527/harmonised/23143594-GCST005527-EFO_0000676-Build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-44" data-line-number="44"></a>
<a class="sourceLine" id="cb3-45" data-line-number="45"><span class="fu">mkdir</span> Rheenen</a>
<a class="sourceLine" id="cb3-46" data-line-number="46"><span class="fu">wget</span> -O Rheenen/raw_rheenen.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/vanRheenenW_27455348_GCST004692/harmonised/27455348-GCST004692-EFO_0000253-build37.f.tsv.gz</a>
<a class="sourceLine" id="cb3-47" data-line-number="47"></a>
<a class="sourceLine" id="cb3-48" data-line-number="48"></a>
<a class="sourceLine" id="cb3-49" data-line-number="49"><span class="co">#Files that need to be harmonized</span></a>
<a class="sourceLine" id="cb3-50" data-line-number="50"></a>
<a class="sourceLine" id="cb3-51" data-line-number="51"><span class="fu">mkdir</span> Jin</a>
<a class="sourceLine" id="cb3-52" data-line-number="52"><span class="fu">wget</span> -O Jin/ ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/JinY_30674883_GCST007111/JinY_PrePMID_GWAS123chr*</a>
<a class="sourceLine" id="cb3-53" data-line-number="53"></a>
<a class="sourceLine" id="cb3-54" data-line-number="54"><span class="fu">mkdir</span> Namjou</a>
<a class="sourceLine" id="cb3-55" data-line-number="55"><span class="fu">wget</span> -O Namjou/raw_namjou.ss ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/NamjouB_31311600_GCST008468/NamjouB_31311600_NAFLD.txt</a>
<a class="sourceLine" id="cb3-56" data-line-number="56"></a>
<a class="sourceLine" id="cb3-57" data-line-number="57"><span class="fu">mkdir</span> Lopez</a>
<a class="sourceLine" id="cb3-58" data-line-number="58"><span class="fu">wget</span> -O Lopez/raw_lopez.ss ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/Lopez-IsacE_31672989_GCST009131/Lopez-Isac_prePMID_META_GWAS_SSc.meta.txt</a>
<a class="sourceLine" id="cb3-59" data-line-number="59"></a>
<a class="sourceLine" id="cb3-60" data-line-number="60"><span class="fu">mkdir</span> Xie</a>
<a class="sourceLine" id="cb3-61" data-line-number="61"><span class="fu">wget</span> -O Xie/raw_xie.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/XieJ_32231244_GCST010004/Trans-ethnic_GWAS_meta-analysis_without_replication.txt.gz</a>
<a class="sourceLine" id="cb3-62" data-line-number="62"></a>
<a class="sourceLine" id="cb3-63" data-line-number="63"><span class="fu">mkdir</span> Mahajan</a>
<a class="sourceLine" id="cb3-64" data-line-number="64"><span class="fu">wget</span> -O Mahajan/raw_mahajan.ss ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/MahajanA_29632382_GCST007515/T2D_TranEthnic.BMIunadjusted.txt</a>
<a class="sourceLine" id="cb3-65" data-line-number="65"></a>
<a class="sourceLine" id="cb3-66" data-line-number="66"><span class="fu">mkdir</span> Shah</a>
<a class="sourceLine" id="cb3-67" data-line-number="67"><span class="fu">wget</span> -O Shah/raw_shah.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/ShahS_31919418_GCST009541/ShahS_31919418_HeartFailure.gz</a>
<a class="sourceLine" id="cb3-68" data-line-number="68"></a>
<a class="sourceLine" id="cb3-69" data-line-number="69"><span class="fu">mkdir</span> Kottgen</a>
<a class="sourceLine" id="cb3-70" data-line-number="70"><span class="fu">wget</span> -O Kottgen/raw_kottgen.ss.gz ftp://ftp.ebi.ac.uk/pub/databases/gwas/ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/KottgenA_23263486_GCST001790/GUGC_MetaAnalysis_Results_Gout.csv.zip</a>
<a class="sourceLine" id="cb3-71" data-line-number="71"></a>
<a class="sourceLine" id="cb3-72" data-line-number="72"><span class="co">#Note there is also data from 23andMe that needs to be processed</span></a></code></pre></div>
<p>This is basically just a series of wget’s to the FTP server of the GWAS catalog. I am also using some files from 23andMe, however these cannot be openly downloaded on GWAS Catalog.</p>
<p>Lastly, a committee member of mine asked me to look into a few pyschiatric related traits, which has summary statistics available on request or within the pyschiatric genomics consortium (<a href="https://www.med.unc.edu/pgc/download-results/" class="uri">https://www.med.unc.edu/pgc/download-results/</a>). While I likely could have followed a similar request approach to recieve a few more sets, the total number of 25 seems like a good stopping point.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw">include_graphics</span>(<span class="st">&quot;./images/gwascatalog.png&quot;</span>)</a></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-3"></span>
<img src="images/gwascatalog.png" alt="test" width="100%" />
<p class="caption">
Figure 4.1: test
</p>
</div>
<p>If you want to manually download just one summary statistic, or you want a more pictoral description of the summary statistics then you can access <a href="https://www.ebi.ac.uk/gwas/downloads/summary-statistics" class="uri">https://www.ebi.ac.uk/gwas/downloads/summary-statistics</a> . There are other GWAS results that are not fully summary statistics, as the motivation of this project is to assess different methods we want the full summary statistics for many SNPs, not just the signifiant ones. Plus, from the full summary statistics we can always reduce down to only a significant-only score.</p>
<div id="prepare-variants" class="section level2">
<h2><span class="header-section-number">4.1</span> Prepare Variants</h2>
<p>As we’ve already QCd the individuals, the other half of the QC is on the variants. That can be easily done by keeping variants that have an imputation INFO score above 0.9, and removing duplicate variants (or variants that are tri-allelic). The INFO score refers to the probability the imputation software has in calling the allele the way it did, we only want high probability variants. We do not want duplicate variants because its tri-allele option slightly increases the probability that it was called or otherwise referred to incorrectly at some point in its analysis. Moreover, the biology is likely more complicating (perhaps not agreeing with the assumptions of the downstream methods), and sorting gets easier.</p>
<p>The other variant level of QC nescessary is on the minor allele frequency and hardy weinberg equilibrium, but we can do that after the white, british group has already be determined. Note that in other analyses this step could be done now, but with file formats as I have (bgen now, plink later), I will wait for this step.</p>
<p>The script used to pull this off is as follows, where the ukb_mfi_all.txt file originates from the UKBB directly.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">#Reduce the full list of RSIDs to those with INFO score greater than 0.9</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="fu">cat</span> ~/athena/ukbiobank/qc/imputed/ukb_mfi_all.txt <span class="kw">|</span> <span class="fu">awk</span> <span class="st">&#39;$8 &gt; 0.9 {print $0}&#39;</span> <span class="op">&gt;</span> common_files/impute_rsids</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#Extract all of the duplicated RSIDs</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="fu">cat</span> ~/athena/ukbiobank/qc/imputed/ukb_mfi_all.txt <span class="kw">|</span> <span class="fu">cut</span> -f2 <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -d <span class="op">&gt;</span> common_files/dup_ids</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#Remove the duplicated RSIDs from the full list of (INFO approved) SNPs</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="fu">cat</span> common_files/impute_rsids <span class="kw">|</span> <span class="fu">fgrep</span> -w -v -f common_files/dup_ids <span class="op">&gt;</span> common_files/temp</a>
<a class="sourceLine" id="cb5-9" data-line-number="9"></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="fu">mv</span> common_files/temp common_files/impute_rsids</a></code></pre></div>
</div>
<div id="standardize-the-summary-statistics" class="section level2">
<h2><span class="header-section-number">4.2</span> Standardize the Summary Statistics</h2>
<p>This is a big step. The underlying motivation is that all of the files need to look the same so downstream coding is far easier. In this process we will convert the base pair positions, or the genome build, so they match the UK Biobank. The can simply be done by matching the rsIDs. We will also remove ambigious SNPs, and then flip the remaining SNPs so they match the UK Biobank strand. The exact procedures, and a few other small changes are described in the script.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(vroom)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">library</span>(stringr)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw">library</span>(bigsnpr)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="kw">options</span>(<span class="dt">warn=</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#impute &lt;- as.data.frame(vroom(&quot;common_files/impute_rsids&quot;, col_names = F))</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8">impute &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;common_files/impute_rsids&quot;</span>, <span class="dt">header =</span> F, <span class="dt">stringsAsFactors =</span> F)</a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="kw">colnames</span>(impute) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;LOC&quot;</span>, <span class="st">&quot;SNP&quot;</span>, <span class="st">&quot;POS&quot;</span>, <span class="st">&quot;A1&quot;</span>, <span class="st">&quot;A2&quot;</span>, <span class="st">&quot;MAF&quot;</span>, <span class="st">&quot;AX&quot;</span>, <span class="st">&quot;INFO&quot;</span>)</a>
<a class="sourceLine" id="cb6-10" data-line-number="10"></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#First remove SNPs where the base is longer than one allele</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#And make sure the SNP is either A, C, G, T</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#Remove ambigous SNPs, those where the bases are actually pairs</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14">impute &lt;-<span class="st"> </span>impute[<span class="kw">nchar</span>(impute<span class="op">$</span>A1) <span class="op">==</span><span class="st"> </span><span class="dv">1</span> <span class="op">&amp;</span><span class="st"> </span><span class="kw">nchar</span>(impute<span class="op">$</span>A2) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>,]</a>
<a class="sourceLine" id="cb6-15" data-line-number="15">impute &lt;-<span class="st"> </span>impute[impute<span class="op">$</span>A1 <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;T&quot;</span>) <span class="op">&amp;</span><span class="st"> </span>impute<span class="op">$</span>A2 <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;T&quot;</span>),]</a>
<a class="sourceLine" id="cb6-16" data-line-number="16">impute &lt;-<span class="st"> </span>impute[<span class="op">!</span>((impute<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;A&quot;</span> <span class="op">&amp;</span><span class="st"> </span>impute<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;T&quot;</span>) <span class="op">|</span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="st">             </span>(impute<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;T&quot;</span> <span class="op">&amp;</span><span class="st"> </span>impute<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;A&quot;</span>) <span class="op">|</span></a>
<a class="sourceLine" id="cb6-18" data-line-number="18"><span class="st">             </span>(impute<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;G&quot;</span> <span class="op">&amp;</span><span class="st"> </span>impute<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;C&quot;</span>) <span class="op">|</span></a>
<a class="sourceLine" id="cb6-19" data-line-number="19"><span class="st">             </span>(impute<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;C&quot;</span> <span class="op">&amp;</span><span class="st"> </span>impute<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;G&quot;</span>)),]</a>
<a class="sourceLine" id="cb6-20" data-line-number="20"></a>
<a class="sourceLine" id="cb6-21" data-line-number="21"></a>
<a class="sourceLine" id="cb6-22" data-line-number="22">adjust_ss &lt;-<span class="st"> </span><span class="cf">function</span>(author){</a>
<a class="sourceLine" id="cb6-23" data-line-number="23"></a>
<a class="sourceLine" id="cb6-24" data-line-number="24">  <span class="co">#Start by reading in the summary statistics file and a parameters file that states the column</span></a>
<a class="sourceLine" id="cb6-25" data-line-number="25">  <span class="co">#names of the desired columns</span></a>
<a class="sourceLine" id="cb6-26" data-line-number="26">  <span class="co">#ss &lt;- as.data.frame(vroom(paste0(author, &quot;/raw_&quot;, tolower(author), &quot;.ss.gz&quot;)))</span></a>
<a class="sourceLine" id="cb6-27" data-line-number="27">  ss &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(author, <span class="st">&quot;/raw_&quot;</span>, <span class="kw">tolower</span>(author), <span class="st">&quot;.ss.gz&quot;</span>), <span class="dt">stringsAsFactors=</span>F, <span class="dt">header =</span> T)</a>
<a class="sourceLine" id="cb6-28" data-line-number="28">  params &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(author, <span class="st">&quot;/parameters&quot;</span>), <span class="dt">stringsAsFactors=</span>F)</a>
<a class="sourceLine" id="cb6-29" data-line-number="29"></a>
<a class="sourceLine" id="cb6-30" data-line-number="30">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo clean notes &gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-31" data-line-number="31">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after read: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-32" data-line-number="32"></a>
<a class="sourceLine" id="cb6-33" data-line-number="33">  <span class="co">#Check if there are missing columns, and fill them in just so things don&#39;t go haywire below</span></a>
<a class="sourceLine" id="cb6-34" data-line-number="34">  missing_cols &lt;-<span class="st"> </span>params[<span class="op">!</span>(params[,<span class="dv">1</span>] <span class="op">%in%</span><span class="st"> </span><span class="kw">colnames</span>(ss)),<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb6-35" data-line-number="35">  <span class="cf">if</span>(<span class="kw">length</span>(missing_cols) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>){</a>
<a class="sourceLine" id="cb6-36" data-line-number="36">    <span class="cf">if</span>(<span class="st">&quot;NO_CHR&quot;</span> <span class="op">%in%</span><span class="st"> </span>missing_cols){</a>
<a class="sourceLine" id="cb6-37" data-line-number="37">      ss<span class="op">$</span>NO_CHR &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb6-38" data-line-number="38">    }</a>
<a class="sourceLine" id="cb6-39" data-line-number="39">    <span class="cf">if</span>(<span class="st">&quot;NO_POS&quot;</span> <span class="op">%in%</span><span class="st"> </span>missing_cols){</a>
<a class="sourceLine" id="cb6-40" data-line-number="40">      ss<span class="op">$</span>NO_POS &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb6-41" data-line-number="41">    }</a>
<a class="sourceLine" id="cb6-42" data-line-number="42">    <span class="cf">if</span>(<span class="st">&quot;NO_SE&quot;</span> <span class="op">%in%</span><span class="st"> </span>missing_cols){</a>
<a class="sourceLine" id="cb6-43" data-line-number="43">      ss<span class="op">$</span>NO_SE &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb6-44" data-line-number="44">    }</a>
<a class="sourceLine" id="cb6-45" data-line-number="45">  }</a>
<a class="sourceLine" id="cb6-46" data-line-number="46"></a>
<a class="sourceLine" id="cb6-47" data-line-number="47"></a>
<a class="sourceLine" id="cb6-48" data-line-number="48">  <span class="co">#Construct a common ss object, with common column names based on the manually curated params</span></a>
<a class="sourceLine" id="cb6-49" data-line-number="49">  ss &lt;-<span class="st">  </span>ss[,<span class="kw">c</span>(<span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">1</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-50" data-line-number="50">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">2</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-51" data-line-number="51">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">3</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-52" data-line-number="52">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">4</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-53" data-line-number="53">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">5</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-54" data-line-number="54">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">6</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-55" data-line-number="55">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">7</span>,<span class="dv">1</span>]),</a>
<a class="sourceLine" id="cb6-56" data-line-number="56">               <span class="kw">which</span>(<span class="kw">colnames</span>(ss) <span class="op">==</span><span class="st"> </span>params[<span class="dv">8</span>,<span class="dv">1</span>]))]</a>
<a class="sourceLine" id="cb6-57" data-line-number="57">  <span class="kw">colnames</span>(ss) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;RSID&quot;</span>, <span class="st">&quot;A1&quot;</span>, <span class="st">&quot;A2&quot;</span>, <span class="st">&quot;SE&quot;</span>, <span class="st">&quot;BETA&quot;</span>, <span class="st">&quot;P&quot;</span>)</a>
<a class="sourceLine" id="cb6-58" data-line-number="58">  ss<span class="op">$</span>A1 &lt;-<span class="st"> </span><span class="kw">toupper</span>(ss<span class="op">$</span>A1)</a>
<a class="sourceLine" id="cb6-59" data-line-number="59">  ss<span class="op">$</span>A2 &lt;-<span class="st"> </span><span class="kw">toupper</span>(ss<span class="op">$</span>A2)</a>
<a class="sourceLine" id="cb6-60" data-line-number="60">  ss &lt;-<span class="st"> </span>ss[<span class="op">!</span><span class="kw">is.na</span>(ss<span class="op">$</span>RSID),]</a>
<a class="sourceLine" id="cb6-61" data-line-number="61">  ss &lt;-<span class="st"> </span>ss[<span class="op">!</span><span class="kw">is.na</span>(ss<span class="op">$</span>BETA),]</a>
<a class="sourceLine" id="cb6-62" data-line-number="62">  ss &lt;-<span class="st"> </span>ss[<span class="op">!</span><span class="kw">is.na</span>(ss<span class="op">$</span>P),]</a>
<a class="sourceLine" id="cb6-63" data-line-number="63"></a>
<a class="sourceLine" id="cb6-64" data-line-number="64">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after remove NA: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-65" data-line-number="65"></a>
<a class="sourceLine" id="cb6-66" data-line-number="66">  <span class="co">#Remove duplicate SNP IDs</span></a>
<a class="sourceLine" id="cb6-67" data-line-number="67">  dup_ids &lt;-<span class="st"> </span>ss<span class="op">$</span>RSID[<span class="kw">duplicated</span>(ss<span class="op">$</span>RSID)]</a>
<a class="sourceLine" id="cb6-68" data-line-number="68">  ss &lt;-<span class="st"> </span>ss[<span class="op">!</span>(ss<span class="op">$</span>RSID <span class="op">%in%</span><span class="st"> </span>dup_ids),]</a>
<a class="sourceLine" id="cb6-69" data-line-number="69"></a>
<a class="sourceLine" id="cb6-70" data-line-number="70">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after remove duplicated IDs: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-71" data-line-number="71"></a>
<a class="sourceLine" id="cb6-72" data-line-number="72">  <span class="co">#First remove SNPs where the base is longer than one allele</span></a>
<a class="sourceLine" id="cb6-73" data-line-number="73">  <span class="co">#And make sure the SNP is either A, C, G, T</span></a>
<a class="sourceLine" id="cb6-74" data-line-number="74">  <span class="co">#Remove ambigous SNPs, those where the bases are actually pairs</span></a>
<a class="sourceLine" id="cb6-75" data-line-number="75">  ss &lt;-<span class="st"> </span>ss[<span class="kw">nchar</span>(ss<span class="op">$</span>A1) <span class="op">==</span><span class="st"> </span><span class="dv">1</span> <span class="op">&amp;</span><span class="st"> </span><span class="kw">nchar</span>(ss<span class="op">$</span>A2) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>,]</a>
<a class="sourceLine" id="cb6-76" data-line-number="76">  ss &lt;-<span class="st"> </span>ss[ss<span class="op">$</span>A1 <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;T&quot;</span>) <span class="op">&amp;</span><span class="st"> </span>ss<span class="op">$</span>A2 <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;T&quot;</span>),]</a>
<a class="sourceLine" id="cb6-77" data-line-number="77">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after remove non ACGT SNPs: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-78" data-line-number="78"></a>
<a class="sourceLine" id="cb6-79" data-line-number="79">  ss &lt;-<span class="st"> </span>ss[<span class="op">!</span>((ss<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;A&quot;</span> <span class="op">&amp;</span><span class="st"> </span>ss<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;T&quot;</span>) <span class="op">|</span></a>
<a class="sourceLine" id="cb6-80" data-line-number="80"><span class="st">             </span>(ss<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;T&quot;</span> <span class="op">&amp;</span><span class="st"> </span>ss<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;A&quot;</span>) <span class="op">|</span></a>
<a class="sourceLine" id="cb6-81" data-line-number="81"><span class="st">             </span>(ss<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;G&quot;</span> <span class="op">&amp;</span><span class="st"> </span>ss<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;C&quot;</span>) <span class="op">|</span></a>
<a class="sourceLine" id="cb6-82" data-line-number="82"><span class="st">             </span>(ss<span class="op">$</span>A1 <span class="op">==</span><span class="st"> &quot;C&quot;</span> <span class="op">&amp;</span><span class="st"> </span>ss<span class="op">$</span>A2 <span class="op">==</span><span class="st"> &quot;G&quot;</span>)),]</a>
<a class="sourceLine" id="cb6-83" data-line-number="83"></a>
<a class="sourceLine" id="cb6-84" data-line-number="84">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after remove ambigous SNPs: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-85" data-line-number="85"></a>
<a class="sourceLine" id="cb6-86" data-line-number="86">  <span class="co">#Make sure the imputation reference and the ss objects have the same number of SNP IDs</span></a>
<a class="sourceLine" id="cb6-87" data-line-number="87">  ss<span class="op">$</span>RSID &lt;-<span class="st"> </span><span class="kw">tolower</span>(ss<span class="op">$</span>RSID)</a>
<a class="sourceLine" id="cb6-88" data-line-number="88">  ss &lt;-<span class="st"> </span>ss[ss<span class="op">$</span>RSID <span class="op">%in%</span><span class="st"> </span>impute<span class="op">$</span>SNP,]</a>
<a class="sourceLine" id="cb6-89" data-line-number="89">  sub_impute &lt;-<span class="st"> </span>impute[impute<span class="op">$</span>SNP <span class="op">%in%</span><span class="st"> </span>ss<span class="op">$</span>RSID,]</a>
<a class="sourceLine" id="cb6-90" data-line-number="90"></a>
<a class="sourceLine" id="cb6-91" data-line-number="91">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after remove SNPs not in the UKBB imputation: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-92" data-line-number="92"></a>
<a class="sourceLine" id="cb6-93" data-line-number="93">  <span class="cf">if</span>(<span class="kw">nrow</span>(ss) <span class="op">!=</span><span class="st"> </span><span class="kw">nrow</span>(sub_impute)){</a>
<a class="sourceLine" id="cb6-94" data-line-number="94">    <span class="kw">print</span>(<span class="st">&quot;SIZE MISMATCH&quot;</span>)</a>
<a class="sourceLine" id="cb6-95" data-line-number="95">  }</a>
<a class="sourceLine" id="cb6-96" data-line-number="96"></a>
<a class="sourceLine" id="cb6-97" data-line-number="97">  <span class="co">#Extract the chromosomes from the imputation reference object</span></a>
<a class="sourceLine" id="cb6-98" data-line-number="98">  <span class="co">#Sometimes the chromosome is not listed for an object</span></a>
<a class="sourceLine" id="cb6-99" data-line-number="99">  <span class="co">#But since the rows seem to be ordered we can impute the chromosome based on the surrounding chromosomes</span></a>
<a class="sourceLine" id="cb6-100" data-line-number="100">  chr &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(sub_impute[,<span class="dv">1</span>], <span class="st">&quot;:&quot;</span>), <span class="cf">function</span>(x) x[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb6-101" data-line-number="101">  <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">as.character</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>)){</a>
<a class="sourceLine" id="cb6-102" data-line-number="102">    <span class="cf">if</span>(i <span class="op">%in%</span><span class="st"> </span>chr){</a>
<a class="sourceLine" id="cb6-103" data-line-number="103">      min_ind &lt;-<span class="st"> </span><span class="kw">min</span>(<span class="kw">which</span>(chr<span class="op">==</span>i))</a>
<a class="sourceLine" id="cb6-104" data-line-number="104">      max_ind &lt;-<span class="st"> </span><span class="kw">max</span>(<span class="kw">which</span>(chr<span class="op">==</span>i))</a>
<a class="sourceLine" id="cb6-105" data-line-number="105">      chr[min_ind<span class="op">:</span>max_ind] &lt;-<span class="st"> </span>i</a>
<a class="sourceLine" id="cb6-106" data-line-number="106">    }</a>
<a class="sourceLine" id="cb6-107" data-line-number="107">  }</a>
<a class="sourceLine" id="cb6-108" data-line-number="108"></a>
<a class="sourceLine" id="cb6-109" data-line-number="109">  <span class="co">#Finish cleaning up the chr, because &quot;X&quot; could be there instead of 23</span></a>
<a class="sourceLine" id="cb6-110" data-line-number="110">  chr[<span class="op">!</span>(chr <span class="op">%in%</span><span class="st"> </span><span class="kw">as.character</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">22</span>))] &lt;-<span class="st"> &quot;23&quot;</span></a>
<a class="sourceLine" id="cb6-111" data-line-number="111">  chr &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(chr)</a>
<a class="sourceLine" id="cb6-112" data-line-number="112"></a>
<a class="sourceLine" id="cb6-113" data-line-number="113">  <span class="co">#Sort the imputation reference and the ss objects so they are the same order</span></a>
<a class="sourceLine" id="cb6-114" data-line-number="114">  <span class="co">#Then assign the imputation chr and pos to the ss object</span></a>
<a class="sourceLine" id="cb6-115" data-line-number="115">  ss &lt;-<span class="st"> </span>ss[<span class="kw">order</span>(ss<span class="op">$</span>RSID)[<span class="kw">rank</span>(sub_impute<span class="op">$</span>SNP)],]</a>
<a class="sourceLine" id="cb6-116" data-line-number="116">  ss &lt;-<span class="st"> </span>ss[<span class="kw">order</span>(chr, sub_impute<span class="op">$</span>POS),]</a>
<a class="sourceLine" id="cb6-117" data-line-number="117">  sub_impute &lt;-<span class="st"> </span>sub_impute[<span class="kw">order</span>(chr, sub_impute<span class="op">$</span>POS),]</a>
<a class="sourceLine" id="cb6-118" data-line-number="118">  chr &lt;-<span class="st"> </span>chr[<span class="kw">order</span>(chr)]</a>
<a class="sourceLine" id="cb6-119" data-line-number="119"></a>
<a class="sourceLine" id="cb6-120" data-line-number="120">  <span class="co">#make the ss object chromosome and position the same as the imputed object</span></a>
<a class="sourceLine" id="cb6-121" data-line-number="121">  ss<span class="op">$</span>CHR &lt;-<span class="st"> </span>chr</a>
<a class="sourceLine" id="cb6-122" data-line-number="122">  ss<span class="op">$</span>BP &lt;-<span class="st"> </span>sub_impute<span class="op">$</span>POS</a>
<a class="sourceLine" id="cb6-123" data-line-number="123"></a>
<a class="sourceLine" id="cb6-124" data-line-number="124"></a>
<a class="sourceLine" id="cb6-125" data-line-number="125">  <span class="co">#If there is a name in the parameters that states CHANGE_BOTH assuming that it&#39;s not BETA but OR</span></a>
<a class="sourceLine" id="cb6-126" data-line-number="126">  <span class="co">#This would just be a simple exponentiation</span></a>
<a class="sourceLine" id="cb6-127" data-line-number="127">  <span class="cf">if</span>(<span class="kw">nrow</span>(params) <span class="op">&gt;</span><span class="st"> </span><span class="dv">8</span>){</a>
<a class="sourceLine" id="cb6-128" data-line-number="128">    <span class="cf">if</span>(params[<span class="dv">9</span>,<span class="dv">1</span>] <span class="op">==</span><span class="st"> &quot;CHANGE_BOTH&quot;</span>){</a>
<a class="sourceLine" id="cb6-129" data-line-number="129">      <span class="co">#will assume that the effect name is for odds ratio</span></a>
<a class="sourceLine" id="cb6-130" data-line-number="130">      <span class="co">#following the walds ratio tests can just switch things over assuming a normal distribution</span></a>
<a class="sourceLine" id="cb6-131" data-line-number="131">      <span class="co">#This may not be exactly what was completed in the published GWAS, but it should be a good approximation</span></a>
<a class="sourceLine" id="cb6-132" data-line-number="132">      <span class="co">#And few methods seem to use the SE column anyway</span></a>
<a class="sourceLine" id="cb6-133" data-line-number="133">      ss<span class="op">$</span>BETA &lt;-<span class="st"> </span><span class="kw">log</span>(ss<span class="op">$</span>BETA)</a>
<a class="sourceLine" id="cb6-134" data-line-number="134">      ss<span class="op">$</span>SE &lt;-<span class="st"> </span><span class="kw">abs</span>(ss<span class="op">$</span>BETA<span class="op">/</span><span class="kw">qnorm</span>(ss<span class="op">$</span>P))</a>
<a class="sourceLine" id="cb6-135" data-line-number="135">    }</a>
<a class="sourceLine" id="cb6-136" data-line-number="136">  }</a>
<a class="sourceLine" id="cb6-137" data-line-number="137"></a>
<a class="sourceLine" id="cb6-138" data-line-number="138"></a>
<a class="sourceLine" id="cb6-139" data-line-number="139">  <span class="co">#use the bignspr function to flip and reverse the summmary statistics</span></a>
<a class="sourceLine" id="cb6-140" data-line-number="140">  <span class="kw">colnames</span>(ss) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;pos&quot;</span>, <span class="st">&quot;rsid&quot;</span>, <span class="st">&quot;a0&quot;</span>, <span class="st">&quot;a1&quot;</span>, <span class="st">&quot;beta_set&quot;</span>, <span class="st">&quot;beta&quot;</span>, <span class="st">&quot;p&quot;</span>)</a>
<a class="sourceLine" id="cb6-141" data-line-number="141">  save_impute_cols &lt;-<span class="st"> </span><span class="kw">colnames</span>(sub_impute)</a>
<a class="sourceLine" id="cb6-142" data-line-number="142">  <span class="kw">colnames</span>(sub_impute) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;loc&quot;</span>, <span class="st">&quot;rsid&quot;</span>, <span class="st">&quot;pos&quot;</span>, <span class="st">&quot;a0&quot;</span>, <span class="st">&quot;a1&quot;</span>, <span class="st">&quot;maf&quot;</span>, <span class="st">&quot;ax&quot;</span>, <span class="st">&quot;info&quot;</span>)</a>
<a class="sourceLine" id="cb6-143" data-line-number="143">  sub_impute<span class="op">$</span>chr &lt;-<span class="st"> </span>ss<span class="op">$</span>chr</a>
<a class="sourceLine" id="cb6-144" data-line-number="144">  ss &lt;-<span class="st"> </span><span class="kw">snp_match</span>(ss, sub_impute)</a>
<a class="sourceLine" id="cb6-145" data-line-number="145">  ss &lt;-<span class="st"> </span>ss[,<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">8</span>)]</a>
<a class="sourceLine" id="cb6-146" data-line-number="146">  <span class="kw">colnames</span>(ss) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CHR&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;RSID&quot;</span>, <span class="st">&quot;A1&quot;</span>, <span class="st">&quot;A2&quot;</span>, <span class="st">&quot;SE&quot;</span>, <span class="st">&quot;BETA&quot;</span>, <span class="st">&quot;P&quot;</span>)</a>
<a class="sourceLine" id="cb6-147" data-line-number="147">  <span class="kw">colnames</span>(sub_impute) &lt;-<span class="st"> </span>save_impute_cols</a>
<a class="sourceLine" id="cb6-148" data-line-number="148">  </a>
<a class="sourceLine" id="cb6-149" data-line-number="149">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after removing SNPs that do not flip: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-150" data-line-number="150"></a>
<a class="sourceLine" id="cb6-151" data-line-number="151"></a>
<a class="sourceLine" id="cb6-152" data-line-number="152">  <span class="co">#Finally add in effective sample size, as defined by PLINK</span></a>
<a class="sourceLine" id="cb6-153" data-line-number="153">  notes &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(author, <span class="st">&quot;/notes&quot;</span>), <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">stringsAsFactors=</span>F)</a>
<a class="sourceLine" id="cb6-154" data-line-number="154">  ss<span class="op">$</span>ESS &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="dv">4</span><span class="op">/</span>(<span class="dv">1</span><span class="op">/</span><span class="kw">as.numeric</span>(notes[<span class="dv">4</span>,<span class="dv">1</span>]) <span class="op">+</span><span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="kw">as.numeric</span>(notes[<span class="dv">5</span>,<span class="dv">1</span>])))</a>
<a class="sourceLine" id="cb6-155" data-line-number="155"></a>
<a class="sourceLine" id="cb6-156" data-line-number="156">  <span class="co">#According to LDPRED-2 it is a good idea to compare to a validation set following:</span></a>
<a class="sourceLine" id="cb6-157" data-line-number="157">  sd_ss &lt;-<span class="st"> </span><span class="dv">2</span><span class="op">/</span>(ss<span class="op">$</span>SE <span class="op">*</span><span class="st"> </span><span class="kw">sqrt</span>(ss<span class="op">$</span>ESS))</a>
<a class="sourceLine" id="cb6-158" data-line-number="158"></a>
<a class="sourceLine" id="cb6-159" data-line-number="159">  <span class="co">#For the validation set I will use summary statistics from the FinnGen Biobank</span></a>
<a class="sourceLine" id="cb6-160" data-line-number="160">  <span class="co">#First I have to go through several steps to line everything up</span></a>
<a class="sourceLine" id="cb6-161" data-line-number="161">  finn_notes &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../finngen_ss/sample_size&quot;</span>, <span class="dt">stringsAsFactors=</span>F)</a>
<a class="sourceLine" id="cb6-162" data-line-number="162">  <span class="cf">if</span>(author <span class="op">%in%</span><span class="st"> </span>finn_notes[,<span class="dv">4</span>]){</a>
<a class="sourceLine" id="cb6-163" data-line-number="163">    finn_ess &lt;-<span class="st"> </span><span class="dv">4</span><span class="op">/</span>((<span class="dv">1</span><span class="op">/</span>finn_notes[finn_notes[,<span class="dv">4</span>] <span class="op">==</span><span class="st"> </span>author,<span class="dv">2</span>]) <span class="op">+</span><span class="st"> </span>(<span class="dv">1</span><span class="op">/</span>finn_notes[finn_notes[,<span class="dv">4</span>] <span class="op">==</span><span class="st"> </span>author,<span class="dv">3</span>]))</a>
<a class="sourceLine" id="cb6-164" data-line-number="164">    finn_pheno &lt;-<span class="st"> </span>finn_notes[finn_notes[,<span class="dv">4</span>] <span class="op">==</span><span class="st"> </span>author,<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb6-165" data-line-number="165"></a>
<a class="sourceLine" id="cb6-166" data-line-number="166">    finn_ss &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(<span class="st">&quot;../finngen_ss/summary_stats_finngen_r3_&quot;</span>, finn_pheno, <span class="st">&quot;.gz&quot;</span>), <span class="dt">header =</span>F, <span class="dt">stringsAsFactors=</span>F, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)  </a>
<a class="sourceLine" id="cb6-167" data-line-number="167">    finn_ss &lt;-<span class="st"> </span>finn_ss[finn_ss[,<span class="dv">5</span>] <span class="op">%in%</span><span class="st"> </span>ss<span class="op">$</span>RSID,]</a>
<a class="sourceLine" id="cb6-168" data-line-number="168">    dup_rs &lt;-<span class="st"> </span>finn_ss[<span class="kw">duplicated</span>(finn_ss[,<span class="dv">5</span>]), <span class="dv">5</span>]</a>
<a class="sourceLine" id="cb6-169" data-line-number="169">    finn_ss &lt;-<span class="st"> </span>finn_ss[<span class="op">!</span>finn_ss[,<span class="dv">5</span>] <span class="op">%in%</span><span class="st"> </span>dup_rs,]</a>
<a class="sourceLine" id="cb6-170" data-line-number="170"></a>
<a class="sourceLine" id="cb6-171" data-line-number="171">    finn_rs &lt;-<span class="st"> </span>finn_ss[,<span class="dv">5</span>]</a>
<a class="sourceLine" id="cb6-172" data-line-number="172">    finn_sd &lt;-<span class="st"> </span>finn_ss[,<span class="dv">9</span>]</a>
<a class="sourceLine" id="cb6-173" data-line-number="173">  </a>
<a class="sourceLine" id="cb6-174" data-line-number="174">    finn_sd &lt;-<span class="st"> </span><span class="kw">c</span>(finn_sd, <span class="kw">rep</span>(<span class="dv">0</span>, (<span class="kw">nrow</span>(ss) <span class="op">-</span><span class="st"> </span><span class="kw">length</span>(finn_rs))))</a>
<a class="sourceLine" id="cb6-175" data-line-number="175">    finn_rs &lt;-<span class="st"> </span><span class="kw">c</span>(finn_rs, ss<span class="op">$</span>RSID[<span class="op">!</span>(ss<span class="op">$</span>RSID <span class="op">%in%</span><span class="st"> </span>finn_rs)])</a>
<a class="sourceLine" id="cb6-176" data-line-number="176">    finn_sd &lt;-<span class="st"> </span>finn_sd[<span class="kw">order</span>(finn_rs)[<span class="kw">rank</span>(ss<span class="op">$</span>RSID)]]</a>
<a class="sourceLine" id="cb6-177" data-line-number="177"></a>
<a class="sourceLine" id="cb6-178" data-line-number="178">    finn_sd &lt;-<span class="st"> </span><span class="dv">2</span><span class="op">/</span>(finn_sd <span class="op">*</span><span class="st"> </span><span class="kw">sqrt</span>(finn_ess))</a>
<a class="sourceLine" id="cb6-179" data-line-number="179"></a>
<a class="sourceLine" id="cb6-180" data-line-number="180">    good_rs &lt;-<span class="st"> </span>ss<span class="op">$</span>RSID[sd_ss <span class="op">&lt;</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>finn_sd <span class="op">&amp;</span><span class="st"> </span>sd_ss <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span> <span class="op">*</span><span class="st"> </span>finn_sd]</a>
<a class="sourceLine" id="cb6-181" data-line-number="181">    good_rs &lt;-<span class="st"> </span><span class="kw">c</span>(good_rs, ss<span class="op">$</span>RSID[finn_sd <span class="op">==</span><span class="st"> </span><span class="ot">Inf</span>])</a>
<a class="sourceLine" id="cb6-182" data-line-number="182">    ss &lt;-<span class="st"> </span>ss[ss<span class="op">$</span>RSID <span class="op">%in%</span><span class="st"> </span>good_rs,]</a>
<a class="sourceLine" id="cb6-183" data-line-number="183">  }</a>
<a class="sourceLine" id="cb6-184" data-line-number="184">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;echo after removing SNPs that do not align with FinnGen: &quot;</span>, <span class="kw">nrow</span>(ss), <span class="st">&quot; &gt;&gt; &quot;</span>, author, <span class="st">&quot;/clean.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-185" data-line-number="185"></a>
<a class="sourceLine" id="cb6-186" data-line-number="186"></a>
<a class="sourceLine" id="cb6-187" data-line-number="187">  <span class="co">#Now write the answer</span></a>
<a class="sourceLine" id="cb6-188" data-line-number="188">  <span class="kw">write.table</span>(ss, <span class="kw">paste0</span>(author, <span class="st">&quot;/clean_&quot;</span>, <span class="kw">tolower</span>(author), <span class="st">&quot;.txt&quot;</span>), <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> T, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb6-189" data-line-number="189">  <span class="kw">system</span>(<span class="kw">paste0</span>(<span class="st">&quot;gzip &quot;</span>, author, <span class="st">&quot;/clean_&quot;</span>, <span class="kw">tolower</span>(author), <span class="st">&quot;.txt&quot;</span>))</a>
<a class="sourceLine" id="cb6-190" data-line-number="190">}</a>
<a class="sourceLine" id="cb6-191" data-line-number="191"></a>
<a class="sourceLine" id="cb6-192" data-line-number="192"></a>
<a class="sourceLine" id="cb6-193" data-line-number="193">all_authors &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;common_files/list_authors&quot;</span>, <span class="dt">stringsAsFactors=</span>F)</a>
<a class="sourceLine" id="cb6-194" data-line-number="194"><span class="cf">for</span>(author <span class="cf">in</span> all_authors[,<span class="dv">1</span>]){</a>
<a class="sourceLine" id="cb6-195" data-line-number="195">  <span class="kw">print</span>(author)</a>
<a class="sourceLine" id="cb6-196" data-line-number="196">  <span class="kw">adjust_ss</span>(author)</a>
<a class="sourceLine" id="cb6-197" data-line-number="197">}</a></code></pre></div>
<p>As you can see at the bottom of the script I am looping over all of the sets of summary statistics, which are organized so each set belongs to its own directory with the name raw_author.ss.gz. Also within the directory are two files, “notes” and “parameters”. The notes files contains some meta-information and the parameters names the column names referring to a different statistic or type of information. Specifically “notes” contains with each new line: author name, disease name, total sample size, case sample size, control sample size, ancestry of GWAS and title of originating publication. Specifically “parameters” contains the chromosome, base pair location, rs prefixed variant ID, effect allele (or first allele), non-effect allele (or second allele), standard error of the effect, effect, and p value. There is an optional ninth line with the term CHANGE_BOTH which indicates the effect is an odds ratio and needs to be converted to a beta values (spanning from negative to positive).</p>
<p>With the script read in for reference, I want to clearly list everything that I am kicking out of the summary statistics. To do this I will call examples from the Phelan summary statistics. The raw download appears as:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">ss &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>T, <span class="dt">nrows =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw">head</span>(ss)</a></code></pre></div>
<pre><code>##          rsid Chromosome Position Effect               Baseline     EAF
## 1 rs145072688          1    10352     TA                      T 0.42360
## 2 rs376342519          1    10616      C CCGCCGTTGCAAAGGCGCGCCG 0.99100
## 3 rs201725126          1    13116      G                      T 0.17040
## 4 rs200579949          1    13118      G                      A 0.17040
## 5  rs75454623          1    14930      G                      A 0.52060
## 6 rs199856693          1    14933      A                      G 0.04797
##   R2_oncoarray overall_OR overall_SE overall_chi2 overall_pvalue
## 1       0.4713  -0.019590    0.02150      0.83020         0.3622
## 2       0.8243   0.013190    0.09657      0.01867         0.8913
## 3       0.3866   0.006643    0.02865      0.05375         0.8167
## 4       0.3866   0.006643    0.02865      0.05375         0.8167
## 5       0.4054   0.022960    0.02127      1.16500         0.2804
## 6       0.3547   0.070590    0.06168      1.31000         0.2524</code></pre>
<ol style="list-style-type: decimal">
<li><p>INFO Score &lt; 0.9 - only want alleles whose identity are certain</p></li>
<li><p>Non-duplicate - want simple base subsititions that are easier to call and process</p></li>
</ol>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">ss &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">629730</span>, <span class="dt">nrows =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw">print</span>(ss)</a></code></pre></div>
<pre><code>##          V1 V2        V3 V4 V5     V6     V7        V8      V9    V10      V11
## 1 rs7517916  1 109370390  A  C 0.2888 0.7717 -0.046220 0.01603 8.3130 0.003937
## 2 rs7517916  1 109370390  G  C 0.3199 0.8600  0.009952 0.01498 0.4416 0.506400</code></pre>
<p>We can see the SNP rs7517916 is tri-allelic, with alleles A, C, G. As this tri-system makes calling more difficult, and strand determination near impossible we discard this SNP altogether.</p>
<ol start="3" style="list-style-type: decimal">
<li>ACGT - similar to above, limiting to the most simple cases of base pair subtitition or single nucleotide polymorphism</li>
</ol>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">ss &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">1</span>, <span class="dt">nrows =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">print</span>(ss)</a></code></pre></div>
<pre><code>##            V1 V2    V3 V4                     V5     V6     V7       V8      V9
## 1 rs145072688  1 10352 TA                      T 0.4236 0.4713 -0.01959 0.02150
## 2 rs376342519  1 10616  C CCGCCGTTGCAAAGGCGCGCCG 0.9910 0.8243  0.01319 0.09657
##       V10    V11
## 1 0.83020 0.3622
## 2 0.01867 0.8913</code></pre>
<p>Here we have two clear examples of SNPs whose alleles are not just A, C, G, T. This paradigm does not work with our strand flipping capabilities (and is harder to call by most sequencing technologies) so we again discard these types of SNPs.</p>
<ol start="4" style="list-style-type: decimal">
<li>Non-ambigous - If the alleles are A and T or C and G, these are called ambigous and will present a problem later on when we try to determine if the standedness of the summary statistics match the strand of our data. This is likely possibly to figure out through allele frequencies, but again, it’s not worth the risk</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">ss &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">9</span>, <span class="dt">nrows =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw">print</span>(ss)</a></code></pre></div>
<pre><code>##            V1 V2    V3   V4 V5     V6    V7      V8      V9     V10    V11
## 1 rs201931625  1 15274 TRUE  A 0.6996 0.432 0.00715 0.02451 0.08508 0.7705</code></pre>
<p>The two alleles described are T and A. As stated, since flipping these alleles leaves us where we started we must again discard this SNP since flipping is impossible.</p>
<ol start="5" style="list-style-type: decimal">
<li><p>Non-X chromosome - As chromosome 23 is very different between men and women (and it’s relatively short), we would likely have to do some special case regressions to set up an accurate model, which makes it easiest just to forget it. Note, no example in Phelan but if it did exist the chromosome column would be X, Y or 23.</p></li>
<li><p>Non-reversable SNPs - We need to flip (or reverse) SNPs such that the allele identities of the summary statistics match the allele identities of the genotypes. We do this by assuming that if they do not line up then we can swap A with T, or C with G (and the converse), and things will work. However, if just one of the SNPs matches after reversing then we have a weird problem that is easiest to resolve by removing the troublesome SNP.</p></li>
</ol>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">sst &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">280</span>, <span class="dt">nrows =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2">ref &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/common_files/impute_rsids&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">5995727</span>, <span class="dt">nrow =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="kw">print</span>(sst)</a></code></pre></div>
<pre><code>##           V1 V2     V3   V4 V5      V6    V7      V8      V9     V10   V11
## 1 rs12184267  1 715265 TRUE  C 0.04354 0.371 0.01388 0.05711 0.05906 0.808</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">print</span>(ref)</a></code></pre></div>
<pre><code>##             V1         V2     V3 V4   V5        V6   V7       V8
## 1 1:715265_C_T rs12184267 715265  C TRUE 0.0366377 TRUE 0.926915</code></pre>
<p>This is an example of reversed SNPs, the A1 of the summary statistic does not match the A1 of the reference data. Therefore we swap the position of the alleles in the summary statistics and similarly reverse the beta or effect value.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">sst &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">375350</span>, <span class="dt">nrows =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2">ref &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/common_files/impute_rsids&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">6358575</span>, <span class="dt">nrow =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb19-3" data-line-number="3"><span class="kw">print</span>(sst)</a></code></pre></div>
<pre><code>##          V1 V2       V3   V4 V5       V6  V7     V8    V9    V10    V11
## 1 rs1390473  1 64804379 TRUE  C 0.001289 -99 0.4462 1.128 0.1566 0.6923</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw">print</span>(ref)</a></code></pre></div>
<pre><code>##               V1        V2       V3 V4 V5         V6 V7       V8
## 1 1:64804379_A_G rs1390473 64804379  A  G 0.00451314  A 0.934127</code></pre>
<p>This is an example of a SNP that needs to be flipped since the summary statistic is on one strand while the complementary base pairs are described in the reference data. In this case the beta or effect value stays the same because the same information is being described in both the summary statistic and reference data, just not using the same language or strand.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">sst &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/Phelan/raw_phelan.ss.gz&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">15414049</span>, <span class="dt">nrows =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2">ref &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;../raw_ss/common_files/impute_rsids&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>F, <span class="dt">skip =</span> <span class="dv">7679970</span>, <span class="dt">nrow =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="kw">print</span>(sst)</a></code></pre></div>
<pre><code>##          V1 V2       V3   V4 V5       V6     V7     V8   V9    V10    V11
## 1 rs8120968 20 61831053 TRUE  G 4.62e-05 0.5318 0.8177 1.37 0.3563 0.5506</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1"><span class="kw">print</span>(ref)</a></code></pre></div>
<pre><code>##                V1        V2       V3 V4 V5         V6 V7       V8
## 1 20:61831053_G_A rs8120968 61831053  G  A 0.00120068  A 0.926685</code></pre>
<p>This is an example of a SNP that is not resolved after either flipping or reversing. In short, the G allele appears to be on the reference stand but the T allele does not. With this confusion we decide to play it safe and discard the SNP altogether.</p>
<ol start="7" style="list-style-type: decimal">
<li>Standard error outliers - Following the supplementary of the method LDPRED2, we can compare standard errors between studies to determine if SNPs are outliers in their effects. For this task I have pulled results from FinnGen Biobank, which while not matching ancestry well with British, does a decent enough job of falling within the European category to work.</li>
</ol>
<p>The remaining steps mostly involve re-aligning, changing column names and the like. Note that we take such a liberal stance on removing SNPs because we assume there is a density of summary statistics such that the removal of any SNP will still leave behind another SNP within a given causal locus that is ultimately important for prediction.</p>
</div>
<div id="heritability-estimation" class="section level2">
<h2><span class="header-section-number">4.3</span> Heritability Estimation</h2>
<p>Many downstream methods require heritability estimates of the trait using the corresponding summary statistics. While many, many heritability estimate methods exist I am limited to those that only require summary statistics, of which I know of 2: Linkage Disequilibirum Score Regression and High-Definition Likelihood (HDL). The first step required is to munge the summary statistics, or further QC them following a munge_sumstats script provided by LDSC. The script for which is simple:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">author</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2">  <span class="va">low_author=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$author</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb27-3" data-line-number="3">  <span class="ex">munge_sumstats</span> --sumstats <span class="va">${author}</span>/clean_<span class="va">${low_author}</span>.txt.gz --N-col ESS --out <span class="va">${author}</span>/<span class="va">${low_author}</span>.munged</a>
<a class="sourceLine" id="cb27-4" data-line-number="4"><span class="kw">done</span></a></code></pre></div>
<p>The next step is the actual heritability estimation, which luckily only requires a simple command line call for each method. We do have to be careful however in making sure all of the columns match up in the arguments. The script therefore looks like:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">author</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="co">#author=Tsoi</span></a>
<a class="sourceLine" id="cb28-3" data-line-number="3">  <span class="va">low_author=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$author</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb28-4" data-line-number="4">  <span class="ex">ldsc</span> --h2 <span class="va">${author}</span>/<span class="va">${low_author}</span>.munged.sumstats.gz --ref-ld-chr ~/athena/refs/eur_w_ld_chr/ --w-ld-chr ~/athena/refs/eur_w_ld_chr/ --out <span class="va">${author}</span>/<span class="va">${low_author}</span>.h2</a>
<a class="sourceLine" id="cb28-5" data-line-number="5"></a>
<a class="sourceLine" id="cb28-6" data-line-number="6">  <span class="ex">Rscript</span> ~/Programs/HDL/HDL.run.R \</a>
<a class="sourceLine" id="cb28-7" data-line-number="7">    gwas.df=<span class="va">${author}</span>/<span class="va">${low_author}</span>.munged.sumstats.gz \</a>
<a class="sourceLine" id="cb28-8" data-line-number="8">    LD.path=~/athena/refs/UKB_imputed_SVD_eigen99_extraction \</a>
<a class="sourceLine" id="cb28-9" data-line-number="9">    output.file=<span class="va">${author}</span>/<span class="va">${low_author}</span>.HDL.h2.Rout</a>
<a class="sourceLine" id="cb28-10" data-line-number="10"><span class="kw">done</span></a></code></pre></div>
<p>We can compile all of the heritability estimates, which are currently sitting in log files, into a single file using a simple bash script. Additionaly, we can include other useful information that may be needed downstream, such as sample size and number of SNPs. The script looks like:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb29-1" data-line-number="1"></a>
<a class="sourceLine" id="cb29-2" data-line-number="2"></a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="bu">echo</span> author,trait,sampe_size,cases,controls,ancestry,snps,ldsc_h2,ldsc_h2_se,hdl_h2,hdl_h2se <span class="op">&gt;</span> meta_stats</a>
<a class="sourceLine" id="cb29-4" data-line-number="4"></a>
<a class="sourceLine" id="cb29-5" data-line-number="5"><span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">author</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb29-6" data-line-number="6">  <span class="va">low_author=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$author</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb29-7" data-line-number="7"></a>
<a class="sourceLine" id="cb29-8" data-line-number="8">  <span class="va">trait=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/notes <span class="kw">|</span> <span class="fu">head</span> -2 <span class="kw">|</span> <span class="fu">tail</span> -1<span class="kw">`</span></a>
<a class="sourceLine" id="cb29-9" data-line-number="9">  <span class="va">samp_size=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/notes <span class="kw">|</span> <span class="fu">head</span> -3 <span class="kw">|</span> <span class="fu">tail</span> -1<span class="kw">`</span></a>
<a class="sourceLine" id="cb29-10" data-line-number="10">  <span class="va">cases=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/notes <span class="kw">|</span> <span class="fu">head</span> -4 <span class="kw">|</span> <span class="fu">tail</span> -1<span class="kw">`</span></a>
<a class="sourceLine" id="cb29-11" data-line-number="11">  <span class="va">controls=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/notes <span class="kw">|</span> <span class="fu">head</span> -5 <span class="kw">|</span> <span class="fu">tail</span> -1<span class="kw">`</span></a>
<a class="sourceLine" id="cb29-12" data-line-number="12">  <span class="va">ancestry=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/notes <span class="kw">|</span> <span class="fu">head</span> -6 <span class="kw">|</span> <span class="fu">tail</span> -1<span class="kw">`</span></a>
<a class="sourceLine" id="cb29-13" data-line-number="13">  <span class="va">snps=</span><span class="kw">`</span><span class="fu">zcat</span> <span class="va">${author}</span>/clean_<span class="va">${low_author}</span>.txt.gz <span class="kw">|</span> <span class="fu">wc</span> -l<span class="kw">`</span></a>
<a class="sourceLine" id="cb29-14" data-line-number="14">  <span class="va">ldsc_h2=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/<span class="va">${low_author}</span>.h2.log <span class="kw">|</span> <span class="fu">fgrep</span> <span class="st">&quot;scale h2&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39;:&#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39; &#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb29-15" data-line-number="15">  <span class="va">ldsc_h2se=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/<span class="va">${low_author}</span>.h2.log <span class="kw">|</span> <span class="fu">fgrep</span> <span class="st">&quot;scale h2&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39;:&#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f3 -d<span class="st">&#39; &#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39;(&#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f1 -d<span class="st">&#39;)&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb29-16" data-line-number="16">  <span class="va">hdl_h2=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/<span class="va">${low_author}</span>.HDL.h2.Rout <span class="kw">|</span> <span class="fu">fgrep</span> Heritability <span class="kw">|</span> <span class="fu">cut</span> -f6 -d<span class="st">&#39; &#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb29-17" data-line-number="17">  <span class="va">hdl_h2se=</span><span class="kw">`</span><span class="fu">cat</span> <span class="va">${author}</span>/<span class="va">${low_author}</span>.HDL.h2.Rout <span class="kw">|</span> <span class="fu">fgrep</span> Heritability <span class="kw">|</span> <span class="fu">cut</span> -f7 -d<span class="st">&#39; &#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39;(&#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f1 -d<span class="st">&#39;)&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb29-18" data-line-number="18"></a>
<a class="sourceLine" id="cb29-19" data-line-number="19">  <span class="bu">echo</span> <span class="va">$author</span>,<span class="va">$trait</span>,<span class="va">$samp_size</span>,<span class="va">$cases</span>,<span class="va">$controls</span>,<span class="va">$ancestry</span>,<span class="va">$snps</span>,<span class="va">$ldsc_h2</span>,<span class="va">$ldsc_h2se</span>,<span class="va">$hdl_h2</span>,<span class="va">$hdl_h2se</span> <span class="op">&gt;&gt;</span> meta_stats</a>
<a class="sourceLine" id="cb29-20" data-line-number="20"><span class="kw">done</span></a></code></pre></div>
<p>The only problem is that we still do not have a single heritability value to use downstream. We can fix this problem by conducting a very simple meta-analysis. We average the estimates inversely weighted by their standard errors. If one of the algorithms did not converge on a set of summary statistics then the other algorithms value stands without any averaging. If both alogortihms failed the value of 0.01, is assumed. This last part is somewhat arbritrary but I figured a low estimate would be best. The script used to perform this basic meta-analysis looks like:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">info &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;meta_stats&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>T)</a>
<a class="sourceLine" id="cb30-2" data-line-number="2"></a>
<a class="sourceLine" id="cb30-3" data-line-number="3">info<span class="op">$</span>hdl_h2[<span class="kw">grepl</span>(<span class="st">&quot;was&quot;</span>, info<span class="op">$</span>hdl_h2)] &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb30-4" data-line-number="4">info<span class="op">$</span>hdl_h2se[<span class="kw">grepl</span>(<span class="st">&quot;estimated&quot;</span>, info<span class="op">$</span>hdl_h2se)] &lt;-<span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb30-5" data-line-number="5"></a>
<a class="sourceLine" id="cb30-6" data-line-number="6">info<span class="op">$</span>hdl_h2 &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(info<span class="op">$</span>hdl_h2)</a>
<a class="sourceLine" id="cb30-7" data-line-number="7">info<span class="op">$</span>hdl_h2se &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(info<span class="op">$</span>hdl_h2se)</a>
<a class="sourceLine" id="cb30-8" data-line-number="8"></a>
<a class="sourceLine" id="cb30-9" data-line-number="9">weights &lt;-<span class="st">  </span><span class="kw">cbind</span>(<span class="dv">1</span><span class="op">/</span>info<span class="op">$</span>ldsc_h2_se, <span class="dv">1</span><span class="op">/</span>info<span class="op">$</span>hdl_h2se)<span class="op">/</span><span class="kw">rowSums</span>(<span class="kw">cbind</span>(<span class="dv">1</span><span class="op">/</span>info<span class="op">$</span>ldsc_h2_se, <span class="dv">1</span><span class="op">/</span>info<span class="op">$</span>hdl_h2se))</a>
<a class="sourceLine" id="cb30-10" data-line-number="10">new_h2 &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">cbind</span>(info<span class="op">$</span>ldsc_h2, info<span class="op">$</span>hdl_h2) <span class="op">*</span><span class="st"> </span>weights)</a>
<a class="sourceLine" id="cb30-11" data-line-number="11"></a>
<a class="sourceLine" id="cb30-12" data-line-number="12">final_h2 &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, <span class="kw">nrow</span>(info))</a>
<a class="sourceLine" id="cb30-13" data-line-number="13">final_h2[info<span class="op">$</span>hdl_h2 <span class="op">==</span><span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> </span>info<span class="op">$</span>ldsc_h2[info<span class="op">$</span>hdl_h2 <span class="op">==</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb30-14" data-line-number="14">final_h2[info<span class="op">$</span>ldsc_h2 <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span> <span class="op">|</span><span class="st"> </span>info<span class="op">$</span>ldsc_h2 <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>] &lt;-<span class="st"> </span>info<span class="op">$</span>hdl_h2[info<span class="op">$</span>ldsc_h2 <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span> <span class="op">|</span><span class="st"> </span>info<span class="op">$</span>ldsc_h2 <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>]</a>
<a class="sourceLine" id="cb30-15" data-line-number="15">final_h2[final_h2 <span class="op">==</span><span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> </span>new_h2[final_h2 <span class="op">==</span><span class="st"> </span><span class="dv">0</span>]</a>
<a class="sourceLine" id="cb30-16" data-line-number="16">final_h2[<span class="kw">is.nan</span>(final_h2)] &lt;-<span class="st"> </span><span class="fl">0.01</span></a>
<a class="sourceLine" id="cb30-17" data-line-number="17"></a>
<a class="sourceLine" id="cb30-18" data-line-number="18">info<span class="op">$</span>h2 &lt;-<span class="st"> </span>final_h2</a>
<a class="sourceLine" id="cb30-19" data-line-number="19"><span class="kw">write.table</span>(info, <span class="st">&quot;meta_stats&quot;</span>, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> T, <span class="dt">quote =</span> F, <span class="dt">sep =</span> <span class="st">&#39;,&#39;</span>)</a></code></pre></div>
</div>
<div id="genetic-correlation-estimation" class="section level2">
<h2><span class="header-section-number">4.4</span> Genetic Correlation Estimation</h2>
<p>One method, smtpred, requires estimation of genetic correlation. Along with the motivation this is just good information to know and plot later on, I will now go through the process of calculating pairwide genetic correlations. Luckily for us this process is quite similar to the heritability estimation above, although I should note the process is quite slower. Because I have two methods that run in two different environments I wrote two scripts this time to execute each.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="co">#double for loop to get pairwise correlations</span></a>
<a class="sourceLine" id="cb31-2" data-line-number="2"><span class="co">#always remember to load the right python environment</span></a>
<a class="sourceLine" id="cb31-3" data-line-number="3"></a>
<a class="sourceLine" id="cb31-4" data-line-number="4"><span class="fu">cat</span> common_files/temp_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">author</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb31-5" data-line-number="5">  <span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">sec_author</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb31-6" data-line-number="6"></a>
<a class="sourceLine" id="cb31-7" data-line-number="7">  <span class="va">low_author=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$author</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb31-8" data-line-number="8">  <span class="va">o_author=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$sec_author</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb31-9" data-line-number="9"></a>
<a class="sourceLine" id="cb31-10" data-line-number="10">  <span class="ex">ldsc</span> --rg <span class="va">${author}</span>/<span class="va">${low_author}</span>.munged.sumstats.gz,<span class="va">${sec_author}</span>/<span class="va">${o_author}</span>.munged.sumstats.gz --ref-ld-chr ~/athena/refs/eur_w_ld_chr/ --w-ld-chr ~/athena/refs/eur_w_ld_chr/ --out gen_corr/ldsc/<span class="va">${low_author}</span>.<span class="va">${o_author}</span>.corr</a>
<a class="sourceLine" id="cb31-11" data-line-number="11"></a>
<a class="sourceLine" id="cb31-12" data-line-number="12">  <span class="kw">done</span></a>
<a class="sourceLine" id="cb31-13" data-line-number="13"><span class="kw">done</span></a></code></pre></div>
<p>and the other is:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="kw">library</span>(HDL)</a>
<a class="sourceLine" id="cb32-2" data-line-number="2"></a>
<a class="sourceLine" id="cb32-3" data-line-number="3">all_authors &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;common_files/list_authors&quot;</span>, <span class="dt">stringsAsFactors =</span> F)</a>
<a class="sourceLine" id="cb32-4" data-line-number="4">gwas_files &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb32-5" data-line-number="5">LD.path &lt;-<span class="st"> &quot;~/athena/refs/UKB_imputed_SVD_eigen99_extraction/&quot;</span></a>
<a class="sourceLine" id="cb32-6" data-line-number="6"></a>
<a class="sourceLine" id="cb32-7" data-line-number="7"><span class="co"># read in all of the authors once up top</span></a>
<a class="sourceLine" id="cb32-8" data-line-number="8">i &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb32-9" data-line-number="9"><span class="cf">for</span>(author <span class="cf">in</span> all_authors[,<span class="dv">1</span>]){</a>
<a class="sourceLine" id="cb32-10" data-line-number="10">  low_author &lt;-<span class="st"> </span><span class="kw">tolower</span>(author)</a>
<a class="sourceLine" id="cb32-11" data-line-number="11">  gwas_files[[i]] &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">paste0</span>(author, <span class="st">&quot;/&quot;</span>, low_author, <span class="st">&quot;.munged.sumstats.gz&quot;</span>), <span class="dt">stringsAsFactors=</span>F, <span class="dt">header=</span>T)</a>
<a class="sourceLine" id="cb32-12" data-line-number="12">  i &lt;-<span class="st"> </span>i <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb32-13" data-line-number="13">}</a>
<a class="sourceLine" id="cb32-14" data-line-number="14"></a>
<a class="sourceLine" id="cb32-15" data-line-number="15"><span class="co"># create a doule for loop to get pairwise correlations</span></a>
<a class="sourceLine" id="cb32-16" data-line-number="16"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">12</span><span class="op">:</span><span class="kw">length</span>(gwas_files)){</a>
<a class="sourceLine" id="cb32-17" data-line-number="17">  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(gwas_files)){</a>
<a class="sourceLine" id="cb32-18" data-line-number="18">    <span class="cf">if</span>(i <span class="op">!=</span><span class="st"> </span>j <span class="op">&amp;</span><span class="st"> </span>(<span class="op">!</span>(i <span class="op">==</span><span class="st"> </span><span class="dv">12</span> <span class="op">&amp;</span><span class="st"> </span>j <span class="op">%in%</span><span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">24</span>))){ <span class="co">#I was having some trouble completing some calculations so I would manually set this change</span></a>
<a class="sourceLine" id="cb32-19" data-line-number="19">      res.HDL &lt;-<span class="st"> </span><span class="kw">HDL.rg</span>(gwas_files[[i]], gwas_files[[j]], LD.path)</a>
<a class="sourceLine" id="cb32-20" data-line-number="20">      <span class="kw">write.table</span>(res.HDL<span class="op">$</span>estimates.df, <span class="kw">paste0</span>(<span class="st">&quot;gen_corr/hdl/&quot;</span>, <span class="kw">tolower</span>(all_authors[i,<span class="dv">1</span>]), <span class="st">&quot;.&quot;</span>, <span class="kw">tolower</span>(all_authors[j,<span class="dv">1</span>]), <span class="st">&quot;.corr.log&quot;</span>), <span class="dt">row.names =</span> T, <span class="dt">col.names =</span> T, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb32-21" data-line-number="21">    }</a>
<a class="sourceLine" id="cb32-22" data-line-number="22">  }</a>
<a class="sourceLine" id="cb32-23" data-line-number="23">}</a></code></pre></div>
<p>Once complete, try to find all of the possible log files that would be made and pull out the key line with the genetic correlation value and its stanadard error, writing to a new long file.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="fu">rm</span> genetic_correlations</a>
<a class="sourceLine" id="cb33-2" data-line-number="2"></a>
<a class="sourceLine" id="cb33-3" data-line-number="3"><span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">fauth1</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb33-4" data-line-number="4">  <span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">fauth2</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb33-5" data-line-number="5"></a>
<a class="sourceLine" id="cb33-6" data-line-number="6">  <span class="va">auth1=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$fauth1</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb33-7" data-line-number="7">  <span class="va">auth2=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$fauth2</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb33-8" data-line-number="8"></a>
<a class="sourceLine" id="cb33-9" data-line-number="9">  <span class="kw">if</span><span class="bu"> [</span> <span class="ot">-e</span> gen_corr/ldsc/<span class="va">${auth1}</span>.<span class="va">${auth2}</span>.corr.log<span class="bu"> ]</span>;<span class="kw">then</span></a>
<a class="sourceLine" id="cb33-10" data-line-number="10">    <span class="va">rg_ldsc=</span><span class="kw">`</span><span class="fu">cat</span> gen_corr/ldsc/<span class="va">${auth1}</span>.<span class="va">${auth2}</span>.corr.log <span class="kw">|</span> <span class="fu">fgrep</span> <span class="st">&quot;Genetic Correlation:&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> -f3 -d<span class="st">&#39; &#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb33-11" data-line-number="11">    <span class="va">se_ldsc=</span><span class="kw">`</span><span class="fu">cat</span> gen_corr/ldsc/<span class="va">${auth1}</span>.<span class="va">${auth2}</span>.corr.log <span class="kw">|</span> <span class="fu">fgrep</span> <span class="st">&quot;Genetic Correlation:&quot;</span> <span class="kw">|</span> <span class="fu">cut</span> -f4 -d<span class="st">&#39; &#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39;(&#39;</span> <span class="kw">|</span> <span class="fu">cut</span> -f1 -d<span class="st">&#39;)&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb33-12" data-line-number="12">  <span class="kw">else</span></a>
<a class="sourceLine" id="cb33-13" data-line-number="13">    <span class="va">rg_ldsc=</span>NA</a>
<a class="sourceLine" id="cb33-14" data-line-number="14">    <span class="va">se_ldsc=</span>NA</a>
<a class="sourceLine" id="cb33-15" data-line-number="15">  <span class="kw">fi</span></a>
<a class="sourceLine" id="cb33-16" data-line-number="16"></a>
<a class="sourceLine" id="cb33-17" data-line-number="17">  <span class="kw">if</span><span class="bu"> [</span> <span class="ot">-e</span> gen_corr/hdl/<span class="va">${auth1}</span>.<span class="va">${auth2}</span>.corr.log<span class="bu"> ]</span>;<span class="kw">then</span></a>
<a class="sourceLine" id="cb33-18" data-line-number="18">    <span class="va">rg_hdl=</span><span class="kw">`</span><span class="fu">cat</span> gen_corr/hdl/<span class="va">${auth1}</span>.<span class="va">${auth2}</span>.corr.log <span class="kw">|</span> <span class="fu">fgrep</span> Correlation <span class="kw">|</span> <span class="fu">cut</span> -f2 -d<span class="st">&#39; &#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb33-19" data-line-number="19">    <span class="va">se_hdl=</span><span class="kw">`</span><span class="fu">cat</span> gen_corr/hdl/<span class="va">${auth1}</span>.<span class="va">${auth2}</span>.corr.log <span class="kw">|</span> <span class="fu">fgrep</span> Correlation <span class="kw">|</span> <span class="fu">cut</span> -f3 -d<span class="st">&#39; &#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb33-20" data-line-number="20">  <span class="kw">else</span></a>
<a class="sourceLine" id="cb33-21" data-line-number="21">    <span class="va">rg_hdl=</span>NA</a>
<a class="sourceLine" id="cb33-22" data-line-number="22">    <span class="va">se_hdl=</span>NA</a>
<a class="sourceLine" id="cb33-23" data-line-number="23">  <span class="kw">fi</span></a>
<a class="sourceLine" id="cb33-24" data-line-number="24"></a>
<a class="sourceLine" id="cb33-25" data-line-number="25">  <span class="bu">echo</span> <span class="va">$auth1</span> <span class="va">$auth2</span> <span class="va">$rg_ldsc</span> <span class="va">$se_ldsc</span> <span class="va">$rg_hdl</span> <span class="va">$se_hdl</span> <span class="op">&gt;&gt;</span> genetic_correlations</a>
<a class="sourceLine" id="cb33-26" data-line-number="26"></a>
<a class="sourceLine" id="cb33-27" data-line-number="27">  <span class="kw">done</span></a>
<a class="sourceLine" id="cb33-28" data-line-number="28"><span class="kw">done</span></a></code></pre></div>
<p>Lastly, we need to go through like before and create just one genetic correlation value from our two estimates. As before, we will average two estimates (if they exist), based on their inverse standard error. Whereas before we were not fine with having an NA value, here that is totally fine. Other non-finite values such as Inf or NaN might also be passed on, that’s fine for now and we will deal with it later on as needed.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">gencorr &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;genetic_correlations&quot;</span>, <span class="dt">stringsAsFactors=</span>F, <span class="dt">fill =</span> T)</a>
<a class="sourceLine" id="cb34-2" data-line-number="2">gencorr<span class="op">$</span>avg &lt;-<span class="st"> </span><span class="ot">NA</span></a>
<a class="sourceLine" id="cb34-3" data-line-number="3">gencorr<span class="op">$</span>se &lt;-<span class="st"> </span><span class="ot">NA</span></a>
<a class="sourceLine" id="cb34-4" data-line-number="4"></a>
<a class="sourceLine" id="cb34-5" data-line-number="5"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(gencorr)){</a>
<a class="sourceLine" id="cb34-6" data-line-number="6"></a>
<a class="sourceLine" id="cb34-7" data-line-number="7">  <span class="cf">if</span>(<span class="kw">is.finite</span>(gencorr[i,<span class="dv">3</span>]) <span class="op">&amp;</span><span class="st"> </span><span class="kw">is.finite</span>(gencorr[i,<span class="dv">5</span>])){</a>
<a class="sourceLine" id="cb34-8" data-line-number="8">    weights &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span><span class="op">/</span>gencorr[i,<span class="dv">4</span>], <span class="dv">1</span><span class="op">/</span>gencorr[i,<span class="dv">6</span>]) <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(<span class="kw">c</span>(<span class="dv">1</span><span class="op">/</span>gencorr[i,<span class="dv">4</span>], <span class="dv">1</span><span class="op">/</span>gencorr[i,<span class="dv">6</span>]))</a>
<a class="sourceLine" id="cb34-9" data-line-number="9">    gencorr<span class="op">$</span>avg[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">c</span>(gencorr[i,<span class="dv">3</span>], gencorr[i,<span class="dv">5</span>]) <span class="op">*</span><span class="st"> </span>weights)</a>
<a class="sourceLine" id="cb34-10" data-line-number="10">    gencorr<span class="op">$</span>se[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">c</span>(gencorr[i,<span class="dv">4</span>], gencorr[i,<span class="dv">6</span>]) <span class="op">*</span><span class="st"> </span>weights)</a>
<a class="sourceLine" id="cb34-11" data-line-number="11"></a>
<a class="sourceLine" id="cb34-12" data-line-number="12">  } <span class="cf">else</span> <span class="cf">if</span>(<span class="kw">is.finite</span>(gencorr[i,<span class="dv">3</span>])){</a>
<a class="sourceLine" id="cb34-13" data-line-number="13">    gencorr<span class="op">$</span>avg[i] &lt;-<span class="st"> </span>gencorr[i,<span class="dv">3</span>]</a>
<a class="sourceLine" id="cb34-14" data-line-number="14">    gencorr<span class="op">$</span>se[i] &lt;-<span class="st"> </span>gencorr[i,<span class="dv">4</span>]</a>
<a class="sourceLine" id="cb34-15" data-line-number="15"></a>
<a class="sourceLine" id="cb34-16" data-line-number="16">  } <span class="cf">else</span> <span class="cf">if</span>(<span class="kw">is.finite</span>(gencorr[i,<span class="dv">5</span>])){</a>
<a class="sourceLine" id="cb34-17" data-line-number="17">    gencorr<span class="op">$</span>avg[i] &lt;-<span class="st"> </span>gencorr[i,<span class="dv">5</span>]</a>
<a class="sourceLine" id="cb34-18" data-line-number="18">    gencorr<span class="op">$</span>se[i] &lt;-<span class="st"> </span>gencorr[i,<span class="dv">6</span>]</a>
<a class="sourceLine" id="cb34-19" data-line-number="19">  } </a>
<a class="sourceLine" id="cb34-20" data-line-number="20"></a>
<a class="sourceLine" id="cb34-21" data-line-number="21">}</a>
<a class="sourceLine" id="cb34-22" data-line-number="22"></a>
<a class="sourceLine" id="cb34-23" data-line-number="23"><span class="kw">write.table</span>(gencorr, <span class="st">&quot;genetic_correlations&quot;</span>, <span class="dt">row.names =</span> F, <span class="dt">col.names =</span> F, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>, <span class="dt">quote =</span> F)</a></code></pre></div>
</div>
<div id="final-changes-and-remarks" class="section level2">
<h2><span class="header-section-number">4.5</span> Final Changes and Remarks</h2>
<p>The last thing to do is to split the polished and cleaned summary statistics into chromosome chunks. This way we can process each chromosome seperately and everything will run much more easily. The script to do this looks like:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="fu">cat</span> common_files/list_authors <span class="kw">|</span> <span class="kw">while</span> <span class="bu">read</span> <span class="va">author</span>;<span class="kw">do</span></a>
<a class="sourceLine" id="cb35-2" data-line-number="2">  <span class="va">low_author=</span><span class="kw">`</span><span class="bu">echo</span> <span class="st">&quot;</span><span class="va">$author</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="fu">tr</span> <span class="st">&#39;[:upper:]&#39;</span> <span class="st">&#39;[:lower:]&#39;</span><span class="kw">`</span></a>
<a class="sourceLine" id="cb35-3" data-line-number="3"></a>
<a class="sourceLine" id="cb35-4" data-line-number="4">  <span class="fu">mkdir</span> <span class="va">${author}</span>/chr_ss</a>
<a class="sourceLine" id="cb35-5" data-line-number="5">  <span class="kw">for</span> <span class="ex">chr</span> in <span class="dt">{1..23}</span><span class="kw">;do</span></a>
<a class="sourceLine" id="cb35-6" data-line-number="6">    <span class="fu">zcat</span> <span class="va">${author}</span>/clean_<span class="va">${low_author}</span>.txt.gz <span class="kw">|</span> <span class="fu">head</span> -1 <span class="op">&gt;</span> <span class="va">${author}</span>/chr_ss/<span class="va">${low_author}</span>_<span class="va">${chr}</span>.ss</a>
<a class="sourceLine" id="cb35-7" data-line-number="7">    <span class="fu">zcat</span> <span class="va">${author}</span>/clean_<span class="va">${low_author}</span>.txt.gz <span class="kw">|</span> <span class="fu">awk</span> -v var=<span class="st">&quot;</span><span class="va">$chr</span><span class="st">&quot;</span> <span class="st">&#39;$1 == var {print $0}&#39;</span> <span class="op">&gt;&gt;</span> <span class="va">${author}</span>/chr_ss/<span class="va">${low_author}</span>_<span class="va">${chr}</span>.ss</a>
<a class="sourceLine" id="cb35-8" data-line-number="8">    <span class="fu">gzip</span> <span class="va">${author}</span>/chr_ss/<span class="va">${low_author}</span>_<span class="va">${chr}</span>.ss</a>
<a class="sourceLine" id="cb35-9" data-line-number="9">  <span class="kw">done</span></a>
<a class="sourceLine" id="cb35-10" data-line-number="10"></a>
<a class="sourceLine" id="cb35-11" data-line-number="11"><span class="kw">done</span></a></code></pre></div>
<p>We could also do this as we process the summary statistics files, but I personally like having the files around like this and don’t mind the memory needed to store them.</p>
<p>We are almost officialy done with QC (remember we still need to look at MAF and HWE). This QC aimed to be on the stringent side to ensure we have good information going into the downstream methods.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="summary-statistics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="adjusting-summary-statistics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
